### ISSUE TITLE
FCall macro and/or Fortran calling convention

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#38872]
In conversation with @vtjnash I mentioned that it might make sense to add a calling convention to `ccall` that implements the GFortran ABI so that we can target different Fortran ABI compilers (see #38836) instead of having it be explicit. 

Alternatively a `@fcall` macro could make sense that converts the arguments to by-ref and handles, characters correctly. 

x-ref: #32870

**Comment by ViralBShah** [comment]
You mean as discussed in #2167 😂

**Comment by simonbyrne** [comment]
It would also need to handle name mangling, though i'm not sure if that is documented anywhere.

**Comment by ViralBShah** [comment]
The name mangling tends to be compiler dependent. I guess you could look at the symbols in a shared library and guess the mangling scheme.

**Comment by simonbyrne** [comment]
I was wondering how CMake did it, and it looks like they literally compile a file with different combinations of names and see what resulting symbols are exported:
https://github.com/Kitware/CMake/blob/fd8eef034168a4008d06ecc87b4522cad9636a28/Modules/FortranCInterface/Detect.cmake

**Comment by ViralBShah** [comment]
@stevengj 's comment in a flexiblas issue is good to link here: https://github.com/mpimd-csc/flexiblas/issues/12#issuecomment-745403270

**Comment by stevengj** [comment]
> I was wondering how CMake did it, and it looks like they literally compile a file with different combinations of names and see what resulting symbols are exported:

Yes, that's also what the [`AC_F77_WRAPPERS` and `AC_FC_WRAPPERS` macros](https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Fortran-Compiler.html) do.  (I actually contributed these to autoconf decades ago.)   Basically, you only have to check a few possibilities to cover all known compilers: upper/lowercase, with or without underscore, and whether to append a second underscore for identifiers containing an underscore.

**Comment by simonbyrne** [comment]
@stevengj do you know how they handle the string issue? Or is that up to the caller?

**Comment by stevengj** [comment]
The string issue is up to the caller AFAIK; it seemed pretty hard to write an autoconf-style compilation check to test for this IIRC, so you are stuck with a table of compilers.

It wasn't a major issue because the only common case of passing strings from C to Fortran was the single-character strings used in the LAPACK/BLAS interfaces, and not passing the length arguments worked with all common compilers for decades AFAIK until gfortran decided it was important to disallow this (sigh).

**Comment by ViralBShah** [comment]
Should we create a label for `ccall` or more generally `ffi`?

**Comment by ViralBShah** [comment]
This literally only comes up in the BLAS dot product:

https://github.com/JuliaLinearAlgebra/libblastrampoline/issues/56
